LZMA on OSX mavericks -- better error messages needed

Error reporting on missing lzma isn't great


<pre>
[jtang@laplace hashstack2 (wip-coreutils)]$ hit fetch http://ftp.gnu.org/gnu/coreutils/coreutils-8.22.tar.xz
Downloading http://ftp.gnu.org/gnu/coreutils/coreutils-8.22.tar.xz...
Downloading 'http://ftp.gnu.org/gnu/coreutils/coreutils-8.22.tar.xz'
[=========================] 100.0% (5.1MB of 5.1MB) 0.775MB/s ETA 0s
[ERROR] Uncaught exception:
Traceback (most recent call last):
  File "/Users/jtang/develop/hashdist/hashdist/cli/main.py", line 173, in help_on_exceptions
    return func(*args, **kw)
  File "/Users/jtang/develop/hashdist/hashdist/cli/main.py", line 147, in command_line_entry_point
    retcode = args.subcommand_handler(ctx, args)
  File "/Users/jtang/develop/hashdist/hashdist/cli/source_cache_cli.py", line 80, in run
    key = store.fetch_archive(args.url, args.type)
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 285, in fetch_archive
    return ArchiveSourceCache(self).fetch_archive(url, type, None)
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 764, in fetch_archive
    return self._download_archive(url, type, expected_hash)
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 768, in _download_archive
    temp_file, hash = self._download_and_hash(url, type)
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 720, in _download_and_hash
    if not create_archive_handler(type).verify(temp_path):
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 851, in verify
    with closing(self.tarfileobj_from_name(filename)) as tarfileobj:
  File "/Users/jtang/develop/hashdist/hashdist/core/source_cache.py", line 912, in tarfileobj_from_name
    import lzma
ImportError: No module named lzma

This exception has not been translated to a human-friendly error message,
please file an issue at https://github.com/hashdist/hashdist/issues pasting
this stack trace.
</pre>

--- a/hashdist/core/source_cache.py
+++ b/hashdist/core/source_cache.py
@@ -900,6 +900,25 @@ class TarballHandler(object):
         return StringIO(archive_data)
 
 
+class TarSubprocessHandler(TarballHandler):
+    """
+    Call external tar
+
+    This handler should only be used as fallback, it lacks some
+    features and/or depends on the vagueries of the host tar.
+    """
+    
+    def verify(self, filename):
+        return True
+
+    def unpack(self, infile, target_dir, hash):
+        self.logger.debug('Calling tar to unpack %s -> %s', infile, target_dir)
+        try:
+            subprocess.check_call(['tar', 'xf', infile.name, '-C', target_dir, '--strip-components=1'])
+        except subprocess.CalledProcessError:
+            raise CorruptSourceCacheError("Archive corrupt and/or cannot be unpacked: '%s'" % filename)
+
+
 class TarGzHandler(TarballHandler):
     type = 'tar.gz'
     exts = ['tar.gz', 'tgz']
@@ -911,6 +930,7 @@ class TarBz2Handler(TarballHandler):
     exts = ['tar.bz2', 'tb2', 'tbz2']
     read_mode = 'r:bz2'
 
+
 class TarXzHandler(TarballHandler):
     type = 'tar.xz'
     exts = ['tar.xz']
@@ -920,21 +940,23 @@ class TarXzHandler(TarballHandler):
     # XXX: so we use lzma module for Python 2 compatibility.
 
     def tarfileobj_from_name(self, filename):
-        try:
-            import lzma
-        except ImportError:
-            self.logger.error("Please install pyliblzma")
         return lzma.LZMAFile(filename)
 
     def tarfileobj_from_data(self, archive_data):
         # XXX: tarfile has built-in 'r:xz' support only in Python 3, so we use lzma module
         # XXX: lzma.open() and lzma.LZMAFile() accept only file names, so we uncompress in memory
-        try:
-            import lzma
-        except ImportError:
-            self.logger.error("Please install pyliblzma")
         from StringIO import StringIO
         return StringIO(lzma.LZMADecompressor().decompress(archive_data))
+
+
+try:
+    import lzma
+except ImportError:
+    # OSX does not include pyliblzma, but OSX tar can unpack tar.xz
+    class TarXzHandler(TarSubprocessHandler):
+        type = 'tar.xz'
+        exts = ['tar.xz']
+
 
 class ZipHandler(object):
     type = 'zip'

